flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocator.java

```diff
@@ -161,158 +173,66 @@ private SharedSlot getOrAllocateSharedSlot(
 					.thenCompose(slotProfile -> slotProvider.allocatePhysicalSlot(
 						new PhysicalSlotRequest(physicalSlotRequestId, slotProfile, slotWillBeOccupiedIndefinitely)))
 					.thenApply(PhysicalSlotRequest.Result::getPhysicalSlot);
-				return new SharedSlot(physicalSlotRequestId, physicalSlotResourceProfile, group, physicalSlotFuture);
+				return new SharedSlot(
+					physicalSlotRequestId,
+					physicalSlotResourceProfile,
+					group,
+					physicalSlotFuture,
+					slotWillBeOccupiedIndefinitely,
+					this::releaseSharedSlot);
 			});
 	}
 
+	private void releaseSharedSlot(ExecutionSlotSharingGroup executionSlotSharingGroup) {
+		SharedSlot slot = sharedSlots.remove(executionSlotSharingGroup);
+		if (slot != null) {
+			slotProvider.cancelSlotRequest(
+				slot.getPhysicalSlotRequestId(),
+				new FlinkException("Slot is being returned from SlotSharingExecutionSlotAllocator."));
+		} else {
+			LOG.debug("There is no slot for ExecutionSlotSharingGroup {} to release", executionSlotSharingGroup);
```

Comment:
True, I think it is some leftover

Reply:
I just found a case that this branch can happen and the latest `checkNotNull(slot)` will fail.
The case is that a task finishes and returns its logical slot which is the only logical slot of a `SharedSlot`. `SharedSlot#returnLogicalSlot()` will be invoked first to release the shared slot and underlying physical slot. The releasing of physical slot, however, could trigger `SharedSlot#release()` which will try to release this shared slot once more.
I think we need either to add RELEASED/ACTIVE states for `SharedSlot` so that it can skip duplicated releasing. Or we need to keep the `if (slot != null)` condition block here. Anyway, the `else` block does not make much sense so I think we can drop it.
WDYT?

Reply:
I think we could try to eliminate the duplicated `SlotSharingExecutionSlotAllocator#releaseSharedSlot` call by adding the `!isEmpty` check at the end of `SharedSlot#release`:
```
if (!isEmpty()) {
	requestedLogicalSlots.clear();
	releaseCallback.accept(executionSlotSharingGroup);
}
```
 and by hardening `SharedSlot#removeLogicalSlotRequest` with:
```
Preconditions.checkState(
	requestedLogicalSlots.removeKeyB(logicalSlotRequestId) != null,
	"Trying to remove a logical slot request which has been either already removed or never created.");
```
wdyt?

Reply:
I think `if (!isEmpty()) {` can work.
But then we are having a implicit assumption that `SharedSlot#release()` will not be invoked before any `SharedSlot#allocateLogicalSlot()` happens, which looks a bit tricky.

Reply:
alright, I added ALLOCATED/RELEASED states and guards for them